File: ios/ReceiptImage.swift
import Foundation
import UIKit

public func receiptImage(config: ReceiptOptions, imageBase64: String) -> [UInt8]
{
  return receiptRender(config: config) { encoder in
    return render(encoder: encoder, imageBase64s: [imageBase64])
  }
}

public func receiptImages(config: ReceiptOptions, imageBase64s: [String])
  -> [UInt8]
{
  return receiptRender(config: config) { encoder in
    return render(encoder: encoder, imageBase64s: imageBase64s)
  }
}

public func receiptPngImages(config: ReceiptOptions, images: [Data]) -> [UInt8]
{
  return receiptRender(config: config) { encoder in
    return renderPng(encoder: encoder, images: images)
  }
}

// MARK: - Optimized functions for direct Data processing
public func receiptImageFromData(config: ReceiptOptions, imageData: Data) throws
  -> [UInt8]
{
  return receiptRender(config: config) { encoder in
    return try renderFromData(encoder: encoder, imageDatas: [imageData])
  }
}

public func receiptImagesFromData(config: ReceiptOptions, imageDatas: [Data])
  throws -> [UInt8]
{
  return receiptRender(config: config) { encoder in
    return try renderFromData(encoder: encoder, imageDatas: imageDatas)
  }
}

// MARK: - Rendering functions
private func render(encoder: EscPosEncoder, imageBase64s: [String]) -> [UInt8] {
  encoder.align("center")
  for img in imageBase64s {
    encoder.image(img)
  }

  encoder.newline()
    .newline()
    .newline()
    .newline()
    .newline()
    .newline()
    .cut()

  return encoder.encode()
}

private func renderPng(encoder: EscPosEncoder, images: [Data]) -> [UInt8] {
  encoder.align("center")
  for img in images {
    encoder.imageData(img)
  }

  encoder.newline()
    .newline()
    .newline()
    .newline()
    .newline()
    .newline()
    .cut()

  return encoder.encode()
}

private func renderFromData(encoder: EscPosEncoder, imageDatas: [Data]) throws
  -> [UInt8]
{
  encoder.align("center")

  for imageData in imageDatas {
    // Convert Data directly to base64 only when needed by the encoder
    let base64String = imageData.base64EncodedString()
    encoder.image(base64String)
  }

  encoder.newline()
    .newline()
    .newline()
    .newline()
    .newline()
    .newline()
    .cut()

  return encoder.encode()
}

private func receiptRender(
  config: ReceiptOptions, render: (EscPosEncoder) throws -> [UInt8]
) -> [UInt8] {
  let modelDefinitions:
    [PrinterModel: (size: Int, codepageCandidates: [String])] = [
      .model58: (58, ["cp866"]),
      .model80: (80, ["cp866"]),
    ]
  let model = modelDefinitions[config.model] ?? (80, ["cp866"])
  let encoder = EscPosEncoder(options: [
    "width": model.size == 58 ? 32 : 46,
    "codepageCandidates": model.codepageCandidates,
  ])
  encoder.initialize()
  encoder.codepage("auto")

  do {
    return try render(encoder)
  } catch {
    return []
  }
}

File: android/src/main/java/expo/modules/escpos/ReceiptImage.kt
package expo.modules.escpos

import expo.modules.escpos.EscPosEncoder.Options

enum class ReceiptModel(val value: String) {
    MODEL_58("58"),
    MODEL_80("80")
}

data class ReceiptConfig(
    val model: ReceiptModel
)

fun receiptImages(config: ReceiptConfig, imageBase64s: List<String>): ByteArray {
    return receiptRender(config) { encoder ->
        renderImages(encoder, imageBase64s)
    }
}

fun receiptPngImages(config: ReceiptConfig, images: List<ByteArray>): ByteArray {
    return receiptRender(config) { encoder ->
        renderPngImages(encoder, images)
    }
}

@OptIn(ExperimentalUnsignedTypes::class)
private fun renderImages(encoder: EscPosEncoder, imageBase64s: List<String>): ByteArray {
    encoder.align("center")

    for (img in imageBase64s) {
        encoder.imageBase64(img)
    }

    // Add space at the end and cut
    encoder.newline()
        .newline()
        .newline()
        .newline()
        .newline()
        .cut()

    return encoder.encode()
}

@OptIn(ExperimentalUnsignedTypes::class)
private fun renderPngImages(encoder: EscPosEncoder, images: List<ByteArray>): ByteArray {
    encoder.align("center")

    for (img in images) {
        encoder.imagePng(img)
    }

    // Add space at the end and cut
    encoder.newline()
        .newline()
        .newline()
        .newline()
        .newline()
        .cut()

    return encoder.encode()
}

private fun receiptRender(config: ReceiptConfig, render: (EscPosEncoder) -> ByteArray): ByteArray {
    val modelDefinitions = mapOf(
        ReceiptModel.MODEL_58 to ModelDefinition(32, listOf("cp866")),
        ReceiptModel.MODEL_80 to ModelDefinition(46, listOf("cp866"))
    )

    val model = modelDefinitions[config.model] ?: ModelDefinition(46, listOf("cp866"))

    val encoder = EscPosEncoder(Options(
        width = model.size,
        codepageCandidates = model.codepageCandidates
    ))

    encoder.initialize()
    encoder.codepage("auto")

    return render(encoder)
}

private data class ModelDefinition(
    val size: Int,
    val codepageCandidates: List<String>
)
